### PR TITLE
initiateWithdraw function

### PR DESCRIPTION
**Pull request by kenchangh**
* initiateWithdraw function that creates a new `Withdrawal` object per address
* `Withdrawal` data structure that takes note of the `round` the withdrawal was initiated in (used to reference `roundPricePerShare` later on)

### CODE REVIEW

File: contracts/storage/OptionsVaultStorage.sol [hunk 1]
```diff
@@ -101,22 +88,21 @@ contract OptionsVaultStorageV3 {
...
 }
 
 // We are following Compound's method of upgrading new contract implementations
 // When we need to add new storage variables, we create a new version of OptionsVaultStorage
 // e.g. OptionsVaultStorageV<versionNumber>, so finally it would look like
 // contract OptionsVaultStorage is OptionsVaultStorageV1, OptionsVaultStorageV2
-contract OptionsVaultStorage is
-    OptionsVaultStorageV1,
-    OptionsVaultStorageV2,
-    OptionsVaultStorageV3
```
Comment (kenchangh) [inline contracts/storage/OptionsVaultStorage.sol:119]: Decided to introduce incompatibility with Ribbon v1, we're going to do a migration of funds from v1 contract to v2 contract

File: test/RibbonThetaVault.ts [hunk 2]
```diff
@@ -1857,225 +1857,128 @@ function behavesLikeRibbonOptionsVault(params: {
...
-
-    //     // Queued withdrawals + 10% of available assets set aside
-    //     assert.equal(
-    //       vaultBalanceBeforeWithdraw.toString(),
-    //       BigNumber.from("110000000000").toString()
-    //     );
-    //   });
-    // });
-
-    // describe("completeScheduledWithdrawal", () => {
```
Comment (kenchangh) [inline test/RibbonThetaVault.ts:1954]: Removing and starting afresh

File: contracts/vaults/RibbonThetaVault.sol [hunk 3]
```diff
@@ -434,6 +435,40 @@ contract RibbonThetaVault is OptionsVaultStorage {
...
+     * @param shares is the number of shares to withdraw
+     */
+    function initiateWithdraw(uint128 shares) external nonReentrant {
+        require(shares > 0, "!shares");
+
+        // This caches the `round` variable used in shareBalances
+        uint16 currentRound = round;
+        Vault.Withdrawal memory withdrawal = withdrawals[msg.sender];
+
+        require(!withdrawal.initiated, "Existing withdraw");
```
Comment (chuddster) [inline contracts/vaults/RibbonThetaVault.sol:449]: so we are constraining it to only one withdrawal per round?

ex:

initiateWithdraw(half of my shares)
10 min later
initiateWithdraw(quarter of my shares)

File: contracts/vaults/RibbonThetaVault.sol [hunk 4]
```diff
@@ -434,6 +435,40 @@ contract RibbonThetaVault is OptionsVaultStorage {
...
+
+        emit InitiateWithdraw(msg.sender, shares, currentRound);
+
+        withdrawals[msg.sender].initiated = true;
+        withdrawals[msg.sender].round = currentRound;
+        withdrawals[msg.sender].shares = shares;
+
+        queuedWithdrawShares = queuedWithdrawShares.add(shares);
+
+        if (shares > heldByVault) {
```
Comment (chuddster) [inline contracts/vaults/RibbonThetaVault.sol:466]: shouldn't it be shares > heldByAccount? if the shares is greater than held by user then we will need to transfer shares from vault to user.

File: contracts/vaults/RibbonThetaVault.sol (hunk 3 above)
Comment (kenchangh) [reply contracts/vaults/RibbonThetaVault.sol:449]: Yep, we're limiting only 1 withdrawal per round. This is really for simplicity

**Review by chuddster — APPROVED** [review]

**Pull request merged**